python/ml4ir/base/data/kfold_relevance_dataset.py

```diff
@@ -0,0 +1,98 @@
+from typing import Optional
+from logging import Logger
+import tensorflow as tf
+
+from ml4ir.base.data.relevance_dataset import RelevanceDataset
+from ml4ir.base.features.feature_config import FeatureConfig
+from ml4ir.base.io.file_io import FileIO
+
+
+class KfoldRelevanceDataset(RelevanceDataset):
+    def __init__(
+            self,
+            data_dir: str,
+            data_format: str,
+            feature_config: FeatureConfig,
+            tfrecord_type: str,
+            file_io: FileIO,
+            max_sequence_size: int = 0,
+            batch_size: int = 128,
+            preprocessing_keys_to_fns: dict = {},
+            train_pcent_split: float = 0.8,
+            val_pcent_split: float = -1,
+            test_pcent_split: float = -1,
+            use_part_files: bool = False,
+            parse_tfrecord: bool = True,
+            logger: Optional[Logger] = None,
+            keep_additional_info: int = 0,
+            non_zero_features_only: int = 0,
+            num_folds: int = 3,
+            include_testset_in_kfold: bool = False,
+            read_data_sets: bool = False
+    ):
+        self.include_testset_in_kfold = include_testset_in_kfold
+        self.num_folds = num_folds
+
+        self.feature_config = feature_config
+        self.max_sequence_size = max_sequence_size
+        self.logger = logger
+        self.data_dir = data_dir
+        self.data_format: str = data_format
+        self.tfrecord_type = tfrecord_type
+        self.batch_size: int = batch_size
+        self.preprocessing_keys_to_fns = preprocessing_keys_to_fns
+        self.file_io = file_io
+
+        self.train_pcent_split: float = train_pcent_split
+        self.val_pcent_split: float = val_pcent_split
+        self.test_pcent_split: float = test_pcent_split
+        self.use_part_files: bool = use_part_files
+
+        self.keep_additional_info = keep_additional_info
+        self.non_zero_features_only = non_zero_features_only
+
+        self.train: Optional[tf.data.TFRecordDataset] = None
+        self.validation: Optional[tf.data.TFRecordDataset] = None
+        self.test: Optional[tf.data.TFRecordDataset] = None
+        if read_data_sets:
+            self.create_dataset(parse_tfrecord)
```

Comment:
@mohazahran Why do you need to do this step again if you already have and plan to use the merged dataset in the method below?

Reply:
The reason for this check is needing to read data from files (creating the dataset) once before iterating through the folds. Then for each fold a new KfoldRelevanceDataset is created (without the need re-read from files). The new object just takes the merged data and slice it according to its fold_idx